WordPress/Classes/Services/AuthenticationService.swift

```diff
@@ -0,0 +1,190 @@
+import AutomatticTracks
+import Foundation
+
+class AuthenticationService {
+
+    static let wpComLoginEndpoint = "https://wordpress.com/wp-login.php"
+
+    // MARK: - Self Hosted
+
+    func loadAuthCookiesForSelfHosted(
+        into cookieJar: CookieJar,
+        loginURL: URL,
+        username: String,
+        password: String,
+        success: @escaping () -> Void,
+        failure: @escaping (Error) -> Void) {
+
+        cookieJar.hasWordPressSelfHostedAuthCookie(for: loginURL, username: username) { hasCookie in
+                guard !hasCookie else {
+                    success()
+                    return
+                }
+
+                self.getAuthCookiesForSelfHosted(loginURL: loginURL, username: username, password: password, success: { cookies in
+                    cookieJar.setCookies(cookies) {
+                        success()
+                    }
+
+                    cookieJar.hasWordPressSelfHostedAuthCookie(for: loginURL, username: username) { hasCookie in
+                        print("Has cookie: \(hasCookie)")
+                    }
+                }) { error in
+                    // Make sure this error scenario isn't silently ignored.
+                    CrashLogging.logError(error)
+
+                    // Even if getting the auth cookies fail, we'll still try to load the URL
+                    // so that the user sees a reasonable error situation on screen.
+                    // We could opt to create a special screen but for now I'd rather users report
+                    // the issue when it happens.
+                    failure(error)
+                }
+        }
+    }
+
+    func getAuthCookiesForSelfHosted(
+        loginURL: URL,
+        username: String,
+        password: String,
+        success: @escaping (_ cookies: [HTTPCookie]) -> Void,
+        failure: @escaping (Error) -> Void) {
+
+        // We don't want these cookies loaded onto all of our requests
+        let session = URLSession(configuration: .ephemeral)
+        var request = URLRequest(url: loginURL)
+
+        request.httpMethod = "POST"
+        request.httpBody = body(withParameters: [
+            "log": username,
+            "pwd": password,
+            "rememberme": "true"
+        ])
+
+        let task = session.dataTask(with: request) { data, response, error in
```

Comment:
It looks like the task code could be factored into its own method accepting a URL, success, and failure blocks if we want to reduce the duplication in the wpcom and selfhosted methods.

Reply:
Been playing a bit with this.

The thing I'd like to avoid is having a single parametrized-method to configure the parameters for the different calls (which is kind of what we had before), since parameter configuration IMHO is specific to each call and not really shared.

With that in mind... how does this look to you?

```swift
private func requestAuthCookies(
        from url: URL,
        headers: [String: String],
        parameters: [String: String],
        success: @escaping (_ cookies: [HTTPCookie]) -> Void,
        failure: @escaping (Error) -> Void) {

        // We don't want these cookies persisted in other sessions
        let session = URLSession(configuration: .ephemeral)
        var request = URLRequest(url: url)

        request.httpMethod = "POST"
        request.httpBody = body(withParameters: parameters)
        
        headers.forEach { (key, value) in
            request.setValue(value, forHTTPHeaderField: key)
        }

        let task = session.dataTask(with: request) { data, response, error in
            if let error = error {
                failure(error)
                return
            }

            // The following code is a bit complicated to read, apologies.
            // We're retrieving all cookies from the "Set-Cookie" header manually, and combining
            // those cookies with the ones from the current session.  The reason behind this is that
            // iOS's URLSession processes the cookies from such header before this callback is executed,
            // whereas OHTTPStubs.framework doesn't (the cookies are left in the header fields of
            // the response).  The only way to combine both is to just add them together here manually.
            //
            // To know if you can remove this, you'll have to test this code live and in our unit tests
            // and compare the session cookies.
            let responseCookies = self.cookies(from: response, loginURL: url)
            let cookies = (session.configuration.httpCookieStorage?.cookies ?? [HTTPCookie]()) + responseCookies
            success(cookies)
        }

        task.resume()
    }
```

Which then can be called like this:

```swift
    func getAuthCookiesForWPCom(
        username: String,
        authToken: String,
        success: @escaping (_ cookies: [HTTPCookie]) -> Void,
        failure: @escaping (Error) -> Void) {
        
        let loginURL = URL(string: AuthenticationService.wpComLoginEndpoint)!
        let headers = [
            "Authorization": "Bearer \(authToken)"
        ]
        let parameters = [
            "log": username,
            "rememberme": "true"
        ]

        requestAuthCookies(
            from: loginURL,
            headers: headers,
            parameters: parameters,
            success: success,
            failure: failure)
    }
```

Reply:
Oh! You went farther than I was thinking. I was just thinking about the task construction.  I agree it would be nice to not over-parameterize things. All I was thinking of was something like

```
func startDataTask(for request: URLRequest,
                        url: URL,
                        success: @escaping (_ cookies: [HTTPCookie]) -> Void,
                        failure: @escaping (Error) -> Void) {
        let session = URLSession(configuration: .ephemeral)
        let task = session.dataTask(with: request) { data, response, error in
            if let error = error {
                failure(error)
                return
            }

            // The following code is a bit complicated to read, apologies.
            // We're retrieving all cookies from the "Set-Cookie" header manually, and combining
            // those cookies with the ones from the current session.  The reason behind this is that
            // iOS's URLSession processes the cookies from such header before this callback is executed,
            // whereas OHTTPStubs.framework doesn't (the cookies are left in the header fields of
            // the response).  The only way to combine both is to just add them together here manually.
            //
            // To know if you can remove this, you'll have to test this code live and in our unit tests
            // and compare the session cookies.
            let responseCookies = self.cookies(from: response, loginURL: url)
            let cookies = (session.configuration.httpCookieStorage?.cookies ?? [HTTPCookie]()) + responseCookies
            success(cookies)
        }

        task.resume()
    }
```

Reply:
I implemented a solution, pls check it out.

Reply:
Oh my, sorry that last message was sent before I read yours! 😄